Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,269 @@
+
+namespace ModuleCat
+
+variable (R S X)
+
+/--
+Let `X` be an `(R,S)`-bimodule. Then `(X ⊗[R] .)` is a functor from the category of `R`-modules
+to the category of `S`-modules.
+-/
+@[simps!]
+noncomputable def tensorFunctor : ModuleCat.{v} R ⥤ ModuleCat.{v} Sᵐᵒᵖ where
+  obj Y := ModuleCat.of Sᵐᵒᵖ <| X ⊗[R] Y
+  map {Y Y'} l :=
+  { __ := l.lTensor _
+    map_smul' := fun s x => x.induction_on (by aesop) (by aesop)
+      fun _ _ (h : l.lTensor _ _ = s • l.lTensor _ _) (h' : l.lTensor _ _ = s • l.lTensor _ _) =>
+        show l.lTensor _ _ = s • l.lTensor _ _ by rw [smul_add, map_add, h, h', map_add, smul_add] }
+  map_id _ := LinearMap.ext fun x => FunLike.congr_fun TensorProduct.map_id x
+  map_comp {_ _ _} l l' := LinearMap.ext fun x => by
+    have eq1 := FunLike.congr_fun (TensorProduct.map_comp LinearMap.id LinearMap.id l' l) x
+    rwa [LinearMap.comp_id] at eq1
+
+/--
+Let `X` be an `(R,S)`-bimodule. Then `(X →ₗ[S] .)` is a functor from the category of `S`-modules
+to the category of `R`-modules.
+-/
+@[simps]
+def homFunctor : ModuleCat.{v} Sᵐᵒᵖ ⥤ ModuleCat.{v} R where
+  obj Z := ModuleCat.of R <| X →ₗ[Sᵐᵒᵖ] Z
+  map {_ _} l :=
+  { toFun := (l ∘ₗ .)
+    map_add' := fun _ _ => LinearMap.ext fun _ => l.map_add _ _
+    map_smul' := fun r L => LinearMap.ext fun x => rfl }
+  map_id Z := LinearMap.ext fun (L : _ →ₗ[Sᵐᵒᵖ] Z) => by aesop
+  map_comp {Z _ _} _ _ := LinearMap.ext fun (L : X →ₗ[Sᵐᵒᵖ] Z) => LinearMap.ext fun x => by aesop
+
+/-- The tensor functor is left adjoint to the hom functor. -/
+noncomputable def tensorHomAdjunction : tensorFunctor R S X ⊣ homFunctor R S X :=
+    Adjunction.mkOfHomEquiv
+    { homEquiv := fun _ _ =>
+      { toFun := hcurry
+        invFun := huncurry
+        left_inv := huncurry_hcurry
+        right_inv := hcurry_huncurry }
+      homEquiv_naturality_left_symm := fun {_ _} _ _ _ => by
+        refine' LinearMap.ext fun x => x.induction_on _ _ _ <;> aesop
+      homEquiv_naturality_right := by aesop }
+
+noncomputable instance : IsLeftAdjoint (tensorFunctor R S X) :=
+⟨_, tensorHomAdjunction _ _ _⟩
+
+noncomputable instance : IsRightAdjoint (homFunctor R S X) :=
+⟨_, tensorHomAdjunction _ _ _⟩
+
+noncomputable instance : Limits.PreservesColimits (tensorFunctor R S X) :=
+Adjunction.leftAdjointPreservesColimits <| tensorHomAdjunction _ _ _
+
+noncomputable instance : Limits.PreservesLimits (homFunctor R S X) :=
+Adjunction.rightAdjointPreservesLimits <| tensorHomAdjunction _ _ _
+
+instance : Functor.PreservesEpimorphisms (tensorFunctor R S X) :=
+inferInstance
+
+end ModuleCat
+
+namespace TensorProduct
+
+variable (R' : Type u) [CommRing R']
+variable {M : Type v} {N : Type v'} [AddCommGroup M] [AddCommGroup N]
+variable [Module R' M] [Module R' N]
+
+open ModuleCat
+
+/--
+Constructing an additive group map from a tensor product by lifting a bi-additive group map that is
+compatible with scalar action.
+-/
+@[simps!]
+noncomputable def toAddCommGroup {C : Type v''} [AddCommGroup C]
```

Comment:
This result is great! I wonder if we could prove it more easily right at the very beginning (before even the module structure on the tensor product), and prove the current `lift` in terms of it?

Reply:
I'll have a PR with that shortly...

Reply:
#8519

Reply:
I actually need this construction more than the adjunction!